quota/etcd/quotaapi/conversions.go

```diff
@@ -0,0 +1,147 @@
+	"google.golang.org/genproto/protobuf/field_mask"
+)
+
+const (
+	statePath           = "state"
+	maxTokensPath       = "max_tokens"
+	sequencingBasedPath = "sequencing_based"
+	timeBasedPath       = "time_based"
+)
+
+var (
+	validPaths = map[string]bool{
+		statePath:           true,
+		maxTokensPath:       true,
+		sequencingBasedPath: true,
+		timeBasedPath:       true,
+	}
+
+	commonMask          = &field_mask.FieldMask{Paths: []string{statePath, maxTokensPath}}
+	sequencingBasedMask = &field_mask.FieldMask{Paths: append(commonMask.Paths, sequencingBasedPath)}
+	timeBasedMask       = &field_mask.FieldMask{Paths: append(commonMask.Paths, timeBasedPath)}
+
+	errBothReplenishmentStrategies = fmt.Errorf("both %q and %q paths specified, please pick a single replenishment strategy", sequencingBasedPath, timeBasedPath)
+)
+
+// validateMask returns nil if mask is a valid Config mask, error otherwise.
+// Paths must match the proto name of the fields (e.g., "time_based", not "TimeBased").
+// If fields belong to a oneof (such as "sequencing_based" and "time_based"), then only one field of
+// the oneof may be specified.
+func validateMask(mask *field_mask.FieldMask) error {
+	sequencingBasedFound := false
+	timeBasedFound := false
+	for _, path := range mask.Paths {
+		switch {
+		case path == sequencingBasedPath:
+			if timeBasedFound {
+				return errBothReplenishmentStrategies
+			}
+			sequencingBasedFound = true
+		case path == timeBasedPath:
+			if sequencingBasedFound {
+				return errBothReplenishmentStrategies
+			}
+			timeBasedFound = true
+		case !validPaths[path]:
+			return fmt.Errorf("invalid field path for Config: %q", path)
+		}
+	}
+	return nil
+}
+
+// applyMask copies the fields specified by mask from src to dest. The mask must be first validated
+// by validateMask(), as unknown paths are simply ignored by applyMask.
+// Paths must match the proto name of the fields (e.g., "time_based", not "TimeBased").
+func applyMask(src *quotapb.Config, dest *storagepb.Config, mask *field_mask.FieldMask) {
+	for _, path := range mask.Paths {
+		switch path {
+		case statePath:
+			dest.State = storagepb.Config_State(storagepb.Config_State_value[src.State.String()])
+		case maxTokensPath:
+			dest.MaxTokens = src.MaxTokens
+		case sequencingBasedPath:
+			if src.GetSequencingBased() == nil {
+				dest.ReplenishmentStrategy = nil
+			} else {
+				dest.ReplenishmentStrategy = &storagepb.Config_SequencingBased{
+					SequencingBased: &storagepb.SequencingBasedStrategy{},
+				}
+			}
+		case timeBasedPath:
+			if tb := src.GetTimeBased(); tb == nil {
+				dest.ReplenishmentStrategy = nil
+			} else {
+				dest.ReplenishmentStrategy = &storagepb.Config_TimeBased{
+					TimeBased: &storagepb.TimeBasedStrategy{
+						TokensToReplenish:        tb.GetTokensToReplenish(),
+						ReplenishIntervalSeconds: tb.GetReplenishIntervalSeconds(),
+					},
+				}
+			}
+		}
+	}
+}
+
+// convertToAPI returns the API representation of a storagepb.Config proto.
```

Comment:
I'm coming in after the fact, but why do we have absolutely identical dual .proto definitions for `Config` and its contents?